// @elastic/react-search-ui/es/containers/ErrorBoundary.js
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _defineProperty from "@babel/runtime/helpers/defineProperty";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

import PropTypes from "prop-types";
import { Component } from "react";
import { withSearch } from "..";
import { ErrorBoundary } from "@elastic/react-search-ui-views";
export var ErrorBoundaryContainer = /*#__PURE__*/function (_Component) {
  _inherits(ErrorBoundaryContainer, _Component);

  var _super = _createSuper(ErrorBoundaryContainer);

  function ErrorBoundaryContainer() {
    _classCallCheck(this, ErrorBoundaryContainer);

    return _super.apply(this, arguments);
  }

  _createClass(ErrorBoundaryContainer, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          children = _this$props.children,
          className = _this$props.className,
          error = _this$props.error,
          view = _this$props.view,
          rest = _objectWithoutProperties(_this$props, ["children", "className", "error", "view"]);

      var View = view || ErrorBoundary;
      return View(_objectSpread({
        className: className,
        children: children,
        error: error
      }, rest));
    }
  }]);

  return ErrorBoundaryContainer;
}(Component);

_defineProperty(ErrorBoundaryContainer, "propTypes", {
  // Props
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
  view: PropTypes.func,
  // State
  error: PropTypes.string.isRequired
});

export default withSearch(function (_ref) {
  var error = _ref.error;
  return {
    error: error
  };
})(ErrorBoundaryContainer);
