// @elastic/react-search-ui/es/containers/PagingInfo.js
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _defineProperty from "@babel/runtime/helpers/defineProperty";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

import PropTypes from "prop-types";
import { Component } from "react";
import { withSearch } from "..";
import { PagingInfo } from "@elastic/react-search-ui-views";
export var PagingInfoContainer = /*#__PURE__*/function (_Component) {
  _inherits(PagingInfoContainer, _Component);

  var _super = _createSuper(PagingInfoContainer);

  function PagingInfoContainer() {
    _classCallCheck(this, PagingInfoContainer);

    return _super.apply(this, arguments);
  }

  _createClass(PagingInfoContainer, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          className = _this$props.className,
          pagingStart = _this$props.pagingStart,
          pagingEnd = _this$props.pagingEnd,
          resultSearchTerm = _this$props.resultSearchTerm,
          totalResults = _this$props.totalResults,
          view = _this$props.view,
          rest = _objectWithoutProperties(_this$props, ["className", "pagingStart", "pagingEnd", "resultSearchTerm", "totalResults", "view"]);

      var View = view || PagingInfo;
      return View(_objectSpread({
        className: className,
        searchTerm: resultSearchTerm,
        start: pagingStart,
        end: pagingEnd,
        totalResults: totalResults
      }, rest));
    }
  }]);

  return PagingInfoContainer;
}(Component);

_defineProperty(PagingInfoContainer, "propTypes", {
  // Props
  className: PropTypes.string,
  view: PropTypes.func,
  // State
  pagingStart: PropTypes.number.isRequired,
  pagingEnd: PropTypes.number.isRequired,
  resultSearchTerm: PropTypes.string.isRequired,
  totalResults: PropTypes.number.isRequired
});

export default withSearch(function (_ref) {
  var pagingStart = _ref.pagingStart,
      pagingEnd = _ref.pagingEnd,
      resultSearchTerm = _ref.resultSearchTerm,
      totalResults = _ref.totalResults;
  return {
    pagingStart: pagingStart,
    pagingEnd: pagingEnd,
    resultSearchTerm: resultSearchTerm,
    totalResults: totalResults
  };
})(PagingInfoContainer);
